File: src/claude-sessions.ts
import { execSync, execFile as execFileCb } from "node:child_process";
import { promisify } from "node:util";
import fs from "node:fs";
import path from "node:path";
import { EventEmitter } from "node:events";

const execFileAsync = promisify(execFileCb);

const CLAUDE_PROJECTS_DIR = path.join(process.env.HOME || "", ".claude", "projects");

export interface ClaudeSession {
  sessionId: string;
  projectPath: string;
  summary: string;
  firstPrompt: string;
  messageCount: number;
  modified: string;
  status: "thinking" | "done" | "idle";
  notified: boolean;
}

interface SessionsIndex {
  version: number;
  entries: Array<{
    sessionId: string;
    fullPath: string;
    summary: string;
    firstPrompt: string;
    messageCount: number;
    modified: string;
    projectPath: string;
  }>;
}

// Track notification state per session
const notificationState = new Map<string, { notified: boolean; viewedAt: Date | null }>();

// Event emitter for session changes
export const sessionEvents = new EventEmitter();

/**
 * Get the status of a Claude session by reading the last line of its jsonl
 */
function getSessionStatus(jsonlPath: string): "thinking" | "done" | "idle" {
  try {
    if (!fs.existsSync(jsonlPath)) {
      return "idle";
    }

    // Read last few KB of the file to find the last complete line
    const stats = fs.statSync(jsonlPath);
    const readSize = Math.min(stats.size, 10000);
    const fd = fs.openSync(jsonlPath, "r");
    const buffer = Buffer.alloc(readSize);
    fs.readSync(fd, buffer, 0, readSize, Math.max(0, stats.size - readSize));
    fs.closeSync(fd);

    const content = buffer.toString("utf-8");
    const lines = content.split("\n").filter(Boolean);

    if (lines.length === 0) {
      return "idle";
    }

    // Parse the last complete JSON line
    const lastLine = lines[lines.length - 1];
    try {
      const msg = JSON.parse(lastLine);

      // "summary" type means conversation turn is complete
      if (msg.type === "summary") {
        return "done";
      }

      // "user" message means Claude is about to respond — but only if recent.
      // If the file hasn't been modified in over 60s, the session was likely
      // interrupted and Claude is not actually thinking.
      if (msg.type === "user") {
        const mtime = stats.mtimeMs;
        const now = Date.now();
        if (now - mtime < 60000) {
          return "thinking";
        }
        return "done";
      }

      // "assistant" message - check if file was recently modified (still streaming)
      if (msg.type === "assistant") {
        const mtime = stats.mtimeMs;
        const now = Date.now();
        // If modified in last 3 seconds, probably still streaming
        if (now - mtime < 3000) {
          return "thinking";
        }
        return "done";
      }

      return "idle";
    } catch {
      return "idle";
    }
  } catch {
    return "idle";
  }
}

/**
 * Find all Claude sessions for a given project path
 */
export function getSessionsForProject(projectPath: string): ClaudeSession[] {
  try {
    // Convert project path to Claude's directory naming convention
    const projectSlug = projectPath.replace(/\//g, "-");
    const projectDir = path.join(CLAUDE_PROJECTS_DIR, projectSlug);

    if (!fs.existsSync(projectDir)) {
      return [];
    }

    const indexPath = path.join(projectDir, "sessions-index.json");

    let entries: Array<{ sessionId: string; fullPath: string; summary: string; firstPrompt: string; messageCount: number; modified: string }>;

    if (fs.existsSync(indexPath)) {
      const index: SessionsIndex = JSON.parse(fs.readFileSync(indexPath, "utf-8"));
      entries = index.entries
        .filter(entry => entry.projectPath === projectPath)
        .map(entry => ({
          sessionId: entry.sessionId,
          fullPath: entry.fullPath,
          summary: entry.summary || "",
          firstPrompt: entry.firstPrompt || "",
          messageCount: entry.messageCount,
          modified: entry.modified,
        }));
    } else {
      // Fallback: scan .jsonl files directly when sessions-index.json is missing
      const files = fs.readdirSync(projectDir).filter(f => f.endsWith(".jsonl"));
      entries = files.map(f => {
        const fullPath = path.join(projectDir, f);
        const sessionId = path.basename(f, ".jsonl");
        const stats = fs.statSync(fullPath);
        return {
          sessionId,
          fullPath,
          summary: "",
          firstPrompt: "",
          messageCount: 0,
          modified: stats.mtime.toISOString(),
        };
      });
    }

    return entries
      .map(entry => {
        // Check and potentially trigger notification
        checkAndNotify(entry.sessionId, entry.fullPath);

        const status = getSessionStatus(entry.fullPath);
        const state = notificationState.get(entry.sessionId) || { notified: false, viewedAt: null };

        return {
          sessionId: entry.sessionId,
          projectPath,
          summary: entry.summary,
          firstPrompt: entry.firstPrompt,
          messageCount: entry.messageCount,
          modified: entry.modified,
          status,
          notified: state.notified,
        };
      })
      .sort((a, b) => new Date(b.modified).getTime() - new Date(a.modified).getTime());
  } catch (err) {
    console.error("Failed to get Claude sessions:", err);
    return [];
  }
}

/**
 * Get the most recent active Claude session for a project
 */
export function getActiveSession(projectPath: string): ClaudeSession | null {
  const sessions = getSessionsForProject(projectPath);

  // Return the most recently modified session
  return sessions[0] || null;
}

/**
 * Mark a session as viewed (clears notification)
 */
export function markSessionViewed(sessionId: string): void {
  const state = notificationState.get(sessionId) || { notified: false, viewedAt: null };
  state.notified = false;
  state.viewedAt = new Date();
  notificationState.set(sessionId, state);
}

/**
 * Get the working directory from a pane's environment
 */
export function getPaneCwd(paneTarget: string): string | null {
  try {
    const output = execSync(`tmux display-message -t "${paneTarget}" -p "#{pane_current_path}"`, {
      encoding: "utf-8",
    }).trim();
    return output || null;
  } catch {
    return null;
  }
}

/**
 * Check if a Claude pane is actively processing by detecting orange thinking status
 * Claude Code thinking indicator: [orange]spinner [salmon]Verbing…[/salmon]
 */
export function isPaneProcessing(paneTarget: string): boolean {
  try {
    // Capture last 10 lines with escape sequences
    const output = execSync(`tmux capture-pane -t "${paneTarget}" -p -e -S -10`, {
      encoding: "utf-8",
    });

    // Orange/salmon color range used by Claude Code thinking status
    const thinkingColor = /\x1b\[38;2;(2[0-3][0-9]);(1[0-5][0-9]);([89][0-9]|1[0-2][0-9])m/;

    // Key: thinking status always has ellipsis "…" (distinguishes from "config (0:0)" etc)
    return thinkingColor.test(output) && output.includes("…");
  } catch {
    return false;
  }
}

// ─── Async variants (non-blocking, for polling paths) ───────────────────────

export async function getPaneCwdAsync(paneTarget: string): Promise<string | null> {
  try {
    const { stdout } = await execFileAsync(
      "tmux",
      ["display-message", "-t", paneTarget, "-p", "#{pane_current_path}"],
      { encoding: "utf-8" },
    );
    return stdout.trim() || null;
  } catch {
    return null;
  }
}

export async function isPaneProcessingAsync(paneTarget: string): Promise<boolean> {
  try {
    const { stdout } = await execFileAsync(
      "tmux",
      ["capture-pane", "-t", paneTarget, "-p", "-e", "-S", "-10"],
      { encoding: "utf-8" },
    );
    const thinkingColor = /\x1b\[38;2;(2[0-3][0-9]);(1[0-5][0-9]);([89][0-9]|1[0-2][0-9])m/;
    return thinkingColor.test(stdout) && stdout.includes("…");
  } catch {
    return false;
  }
}

// Track previous status for change detection
const previousStatus = new Map<string, string>();

/**
 * Check if a session should trigger a notification
 * This is called both on file change and when first accessing a session
 */
function checkAndNotify(sessionId: string, fullPath: string): void {
  const status = getSessionStatus(fullPath);
  const prevStatus = previousStatus.get(sessionId);
  const state = notificationState.get(sessionId) || { notified: false, viewedAt: null };

  // Reset viewedAt when a new turn starts (status leaves "done")
  if (prevStatus === "done" && status !== "done") {
    state.viewedAt = null;
  }

  // Detect transition to "done"
  if (prevStatus === "thinking" && status === "done") {
    console.log(`Claude session ${sessionId} completed`);
    state.notified = true;
    notificationState.set(sessionId, state);
    sessionEvents.emit("completed", sessionId);
  }

  // Also notify if session is "done" and hasn't been notified yet
  // (handles case where server starts after Claude finished)
  // But don't re-notify if user already viewed it while in "done" state
  if (status === "done" && !state.notified && !state.viewedAt) {
    console.log(`Claude session ${sessionId} needs attention (done)`);
    state.notified = true;
    notificationState.set(sessionId, state);
  }

  previousStatus.set(sessionId, status);
}

/**
 * Start watching for Claude session changes
 */
export function startWatching(): void {
  if (!fs.existsSync(CLAUDE_PROJECTS_DIR)) {
    console.log("Claude projects directory not found, skipping session watching");
    return;
  }

  console.log("Watching Claude sessions at:", CLAUDE_PROJECTS_DIR);

  // Watch for file changes
  fs.watch(CLAUDE_PROJECTS_DIR, { recursive: true }, (eventType, filename) => {
    if (!filename || !filename.endsWith(".jsonl")) {
      return;
    }

    // Extract session ID from filename
    const sessionId = path.basename(filename, ".jsonl");
    const fullPath = path.join(CLAUDE_PROJECTS_DIR, filename);

    checkAndNotify(sessionId, fullPath);
  });
}

File: src-tauri/src/claude_sessions.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::io::{Read, Seek, SeekFrom};
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use std::time::SystemTime;

static STATE: once_cell::sync::Lazy<Mutex<ClaudeState>> =
    once_cell::sync::Lazy::new(|| Mutex::new(ClaudeState::default()));

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClaudeSession {
    pub session_id: String,
    pub summary: String,
    pub status: String, // "thinking" | "done" | "idle"
    pub notified: bool,
}

#[derive(Default)]
struct ClaudeState {
    /// notification state per session: (notified, viewed_at)
    notification: HashMap<String, (bool, Option<SystemTime>)>,
    /// previous status for change detection
    previous_status: HashMap<String, String>,
}

fn claude_projects_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_default()
        .join(".claude")
        .join("projects")
}

/// Read the status of a Claude session from its JSONL file
fn get_session_status(jsonl_path: &Path) -> &'static str {
    let meta = match fs::metadata(jsonl_path) {
        Ok(m) => m,
        Err(_) => return "idle",
    };

    let file_size = meta.len();
    if file_size == 0 {
        return "idle";
    }

    // Read last 10KB of file
    let read_size = file_size.min(10000) as usize;
    let mut file = match fs::File::open(jsonl_path) {
        Ok(f) => f,
        Err(_) => return "idle",
    };

    if file_size > read_size as u64 {
        let _ = file.seek(SeekFrom::Start(file_size - read_size as u64));
    }

    let mut buffer = vec![0u8; read_size];
    let bytes_read = match file.read(&mut buffer) {
        Ok(n) => n,
        Err(_) => return "idle",
    };
    buffer.truncate(bytes_read);

    let content = String::from_utf8_lossy(&buffer);
    let lines: Vec<&str> = content.lines().filter(|l| !l.is_empty()).collect();
    if lines.is_empty() {
        return "idle";
    }

    let last_line = lines[lines.len() - 1];
    let msg: serde_json::Value = match serde_json::from_str(last_line) {
        Ok(v) => v,
        Err(_) => return "idle",
    };

    let msg_type = msg.get("type").and_then(|v| v.as_str()).unwrap_or("");

    let mtime = meta
        .modified()
        .ok()
        .and_then(|t| t.elapsed().ok())
        .map(|d| d.as_millis())
        .unwrap_or(u128::MAX);

    match msg_type {
        "summary" => "done",
        "user" => {
            if mtime < 60_000 {
                "thinking"
            } else {
                "done"
            }
        }
        "assistant" => {
            if mtime < 3_000 {
                "thinking"
            } else {
                "done"
            }
        }
        _ => "idle",
    }
}

/// Get all Claude sessions for a project path
pub fn get_sessions_for_project(project_path: &str) -> Vec<ClaudeSession> {
    let project_slug = project_path.replace('/', "-");
    let project_dir = claude_projects_dir().join(&project_slug);

    if !project_dir.exists() {
        return vec![];
    }

    // Try sessions-index.json first
    let index_path = project_dir.join("sessions-index.json");

    #[derive(Deserialize)]
    struct IndexEntry {
        #[serde(rename = "sessionId")]
        session_id: String,
        #[serde(rename = "fullPath")]
        full_path: String,
        summary: Option<String>,
        #[serde(rename = "projectPath")]
        project_path: Option<String>,
    }

    #[derive(Deserialize)]
    struct SessionsIndex {
        entries: Vec<IndexEntry>,
    }

    let entries: Vec<(String, PathBuf, String)> = if index_path.exists() {
        match fs::read_to_string(&index_path).ok().and_then(|s| {
            serde_json::from_str::<SessionsIndex>(&s).ok()
        }) {
            Some(index) => index
                .entries
                .into_iter()
                .filter(|e| {
                    e.project_path
                        .as_deref()
                        .map(|p| p == project_path)
                        .unwrap_or(true)
                })
                .map(|e| {
                    (
                        e.session_id,
                        PathBuf::from(e.full_path),
                        e.summary.unwrap_or_default(),
                    )
                })
                .collect(),
            None => vec![],
        }
    } else {
        // Fallback: scan .jsonl files directly
        match fs::read_dir(&project_dir) {
            Ok(entries) => entries
                .filter_map(|e| e.ok())
                .filter(|e| {
                    e.path()
                        .extension()
                        .map(|ext| ext == "jsonl")
                        .unwrap_or(false)
                })
                .map(|e| {
                    let path = e.path();
                    let session_id = path
                        .file_stem()
                        .unwrap_or_default()
                        .to_string_lossy()
                        .to_string();
                    (session_id, path, String::new())
                })
                .collect(),
            Err(_) => vec![],
        }
    };

    let mut state = STATE.lock().unwrap();
    let mut results: Vec<ClaudeSession> = entries
        .into_iter()
        .map(|(session_id, full_path, summary)| {
            check_and_notify(&mut state, &session_id, &full_path);
            let status = get_session_status(&full_path).to_string();
            let (notified, _) = state
                .notification
                .get(&session_id)
                .copied()
                .unwrap_or((false, None));

            ClaudeSession {
                session_id,
                summary,
                status,
                notified,
            }
        })
        .collect();

    // Sort by most recent first (we can't easily sort by modified time here,
    // but the index entries are already in a reasonable order)
    results.reverse();
    results
}

/// Get the most recent active Claude session for a project
pub fn get_active_session(project_path: &str) -> Option<ClaudeSession> {
    let sessions = get_sessions_for_project(project_path);
    sessions.into_iter().next()
}

/// Mark a session as viewed (clears notification)
pub fn mark_session_viewed(session_id: &str) {
    let mut state = STATE.lock().unwrap();
    state
        .notification
        .insert(session_id.to_string(), (false, Some(SystemTime::now())));
}

/// Check if a session should trigger a notification
fn check_and_notify(state: &mut ClaudeState, session_id: &str, full_path: &Path) {
    let status = get_session_status(full_path).to_string();
    let prev_status = state.previous_status.get(session_id).cloned();
    let (mut notified, mut viewed_at) = state
        .notification
        .get(session_id)
        .copied()
        .unwrap_or((false, None));

    // Reset viewedAt when a new turn starts
    if prev_status.as_deref() == Some("done") && status != "done" {
        viewed_at = None;
    }

    // Detect thinking → done transition
    if prev_status.as_deref() == Some("thinking") && status == "done" {
        log::info!("Claude session {} completed", session_id);
        notified = true;
    }

    // Also notify if done and hasn't been notified yet (and not viewed)
    if status == "done" && !notified && viewed_at.is_none() {
        log::info!("Claude session {} needs attention (done)", session_id);
        notified = true;
    }

    state
        .notification
        .insert(session_id.to_string(), (notified, viewed_at));
    state
        .previous_status
        .insert(session_id.to_string(), status);
}

/// Start watching Claude session files for changes
pub async fn start_watching(app_handle: tauri::AppHandle) {
    let projects_dir = claude_projects_dir();
    if !projects_dir.exists() {
        log::info!("Claude projects directory not found, skipping session watching");
        return;
    }

    log::info!("Watching Claude sessions at: {:?}", projects_dir);

    // Use notify crate for file watching
    use notify::{Config, Event, RecommendedWatcher, RecursiveMode, Watcher};
    use std::sync::mpsc;

    let (tx, rx) = mpsc::channel::<notify::Result<Event>>();

    let mut watcher = match RecommendedWatcher::new(tx, Config::default()) {
        Ok(w) => w,
        Err(e) => {
            log::warn!("Failed to create file watcher: {}", e);
            return;
        }
    };

    if let Err(e) = watcher.watch(&projects_dir, RecursiveMode::Recursive) {
        log::warn!("Failed to watch Claude projects dir: {}", e);
        return;
    }

    // Keep watcher alive by moving it into the task
    let _watcher = watcher;
    let _app_handle = app_handle;

    // Process file change events
    tokio::task::spawn_blocking(move || {
        for result in rx {
            if let Ok(event) = result {
                for path in &event.paths {
                    if let Some(ext) = path.extension() {
                        if ext == "jsonl" {
                            let session_id = path
                                .file_stem()
                                .unwrap_or_default()
                                .to_string_lossy()
                                .to_string();
                            let mut state = STATE.lock().unwrap();
                            check_and_notify(&mut state, &session_id, path);
                        }
                    }
                }
            }
        }
    });
}
